### PR TITLE
FIX: allow raw_fif to parse brainvision events

### PR DESCRIPTION
**Pull request by massich**
This PR fixes this (see discussion in #6326):
```py
raw = read_raw_brainvision(...)
raw.save(fname)
raw_read = read_raw_fif(fname)
events_from_annotations(raw_read, event_id='auto') == events_from_annotations(raw, event_id='auto') # now they are equal!!
```

### CODE REVIEW

File: mne/annotations.py [hunk 1]
```diff
@@ -817,10 +817,18 @@ def _select_annotations_based_on_description(descriptions, event_id, regexp):
...
 
 def _check_event_id(event_id, raw):
+    from mne.io.brainvision.brainvision import _BVEventParser
+    from mne.io.brainvision.brainvision import _check_bv_annot
+
     if event_id is None:
         return _DefaultEventParser()
     elif event_id == 'auto':
-        return getattr(raw, '_get_auto_event_id', _DefaultEventParser)()
+        candidate = getattr(raw, '_get_auto_event_id', _DefaultEventParser)
```
Comment (massich) [inline mne/annotations.py]: I don't think that using a `@classmethod` as a dispatcher makes sense anymore.

@larsoner I would like to see your comments on that. 
('cos I guess that if we get rid of this we'll have to deprecate the class method don't we?)

File: mne/annotations.py [hunk 2]
```diff
@@ -817,10 +817,18 @@ def _select_annotations_based_on_description(descriptions, event_id, regexp):
 
 
 def _check_event_id(event_id, raw):
+    from mne.io.brainvision.brainvision import _BVEventParser
+    from mne.io.brainvision.brainvision import _check_bv_annot
```
Comment (massich) [inline mne/annotations.py]: I'm aware of the imports (I'll push a commit to change them)

File: mne/io/brainvision/tests/test_brainvision.py [hunk 3]
```diff
@@ -519,4 +520,18 @@ def test_automatic_vmrk_sfreq_recovery():
                        read_annotations(vmrk_path, sfreq=1000.0))
 
 
+@testing.requires_testing_data
+def test_event_id_stability_when_save_and_fif_reload(tmpdir):
+    """Test load events from brainvision annotations when read_raw_fif."""
+    fname = str(Path(tmpdir, 'bv-raw.fif'))
```
Comment (larsoner) [inline mne/io/brainvision/tests/test_brainvision.py]: why `Path` and not just `fname = op.join(str(tmpdir), 'bv-raw.fif')` like we do everywhere else?

File: mne/io/brainvision/brainvision.py [hunk 4]
```diff
@@ -849,3 +850,10 @@ def __call__(self, description):
             code = (super(_BVEventParser, self)
                     .__call__(description, offset=_OTHER_OFFSET))
         return code
+
+
+def _check_bv_annot(descriptions):
+    markers_basename = set([dd.rstrip('0123456789 ') for dd in descriptions])
+    bv_markers = set(chain(_BV_EVENT_IO_OFFSETS.keys(),
+                           _OTHER_ACCEPTED_MARKERS.keys()))
```
Comment (larsoner) [inline mne/io/brainvision/brainvision.py]: rather than `chain` why not just `set(a) + set(b)`?

File: mne/annotations.py [hunk 5]
```diff
@@ -817,10 +817,18 @@ def _select_annotations_based_on_description(descriptions, event_id, regexp):
...
+
     if event_id is None:
         return _DefaultEventParser()
     elif event_id == 'auto':
-        return getattr(raw, '_get_auto_event_id', _DefaultEventParser)()
+        candidate = getattr(raw, '_get_auto_event_id', _DefaultEventParser)
+        if candidate == _DefaultEventParser and \
+           _check_bv_annot(raw.annotations.description):
+            return _BVEventParser()
```
Comment (larsoner) [inline mne/annotations.py]: There was an additional qualifier @agramfort wanted which was that `raw` be an instance of `RawFIF` (or I think it should also include `RawArray`)

File: mne/annotations.py [hunk 6]
```diff
@@ -817,10 +817,18 @@ def _select_annotations_based_on_description(descriptions, event_id, regexp):
...
+    from mne.io.brainvision.brainvision import _BVEventParser
+    from mne.io.brainvision.brainvision import _check_bv_annot
+
     if event_id is None:
         return _DefaultEventParser()
     elif event_id == 'auto':
-        return getattr(raw, '_get_auto_event_id', _DefaultEventParser)()
+        candidate = getattr(raw, '_get_auto_event_id', _DefaultEventParser)
+        if candidate == _DefaultEventParser and \
+           _check_bv_annot(raw.annotations.description):
```
Comment (larsoner) [inline mne/annotations.py]: typically an extra 4-space indentation is used on these to make it clear it's part of the conditional of the `if` and not the first statement (helps readability, also PEP8 might complain)

File: mne/annotations.py (hunk 1 above)
Comment (larsoner) [reply mne/annotations.py]: the class method is private, no need for dep cycle, feel free to simplify

File: mne/io/brainvision/tests/test_brainvision.py (hunk 3 above)
Comment (massich) [reply mne/io/brainvision/tests/test_brainvision.py]: yes. I should do it for consistency here. I was tired of `str(tmpdir)` and I thought that using Path would solve it. But the problem is with the reader rather than with `tmpdir`. 

I think I'll open a PR to add Path to the readers. Path is nice.
https://treyhunner.com/2018/12/why-you-should-be-using-pathlib/

But here I just forgot to revert.

File: mne/io/brainvision/brainvision.py (hunk 4 above)
Comment (massich) [reply mne/io/brainvision/brainvision.py]: That was my first relfex, but it was not working.
```sh
xx = set(_BV_EVENT_IO_OFFSETS.keys()) + set(_OTHER_ACCEPTED_MARKERS.keys())
*** TypeError: unsupported operand type(s) for +: 'set' and 'set'
```

Comment (massich) [reply mne/io/brainvision/brainvision.py]:
```sh
In [12]: set({'a':'a', 'b':'b'}.keys()) + set({'c':'c','d':'d'}.keys())                                                                           
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-12-c637c393037a> in <module>
----> 1 set({'a':'a', 'b':'b'}.keys()) + set({'c':'c','d':'d'}.keys())

TypeError: unsupported operand type(s) for +: 'set' and 'set'
```

File: mne/annotations.py (hunk 5 above)
Comment (massich) [reply mne/annotations.py]: That's why I started by implementing this inside a class method of `Raw`, but then `getattr` does nothing because raw is a subclass and its always present

**Comment by massich** [comment]
The other concern I have with this implementation is that right now is strict. Only if all the descriptors belong to 'brainvision' it would be considered brainvision. If you have some extra markers, then it would be considered not brainvision. Which to me is fine. If you have been messing up the markers, saving and reloading etc.. you should be knowing what you are doing.

**Comment by massich** [comment]
I would implement it like this:
```py
def _check_event_id(event_id, raw):
    from mne.io.brainvision.brainvision import _BVEventParser
    from mne.io.brainvision.brainvision import _check_bv_annot

    if event_id is None:
        return _DefaultEventParser()
    elif event_id == 'auto':
        if _check_bv_annot(raw.annotations.description):
            return _BVEventParser()
        else:
            return _DefaultEventParser()
    elif callable(event_id) or isinstance(event_id, dict):
        return event_id
    else:
        raise ValueError('Invalid input event_id')
```

File: mne/annotations.py (hunk 5 above)
Comment (massich) [reply mne/annotations.py]: up to some point, this part was easy: just get read of the default and the base class would already give you `_DefaultEventParser`.

But the method itself was weird:
```
    @classmethod
    def _get_auto_event_id(cls):
        """Return default ``event_id`` behavior unless it matches Brainvision."""
        if _check_annotation_descriptions:  # self is not present
            return _DefaultEventParser()
        else:
            return _BVEventParser()  # return a free helper function of a class that inherates from RaW.
```

returning `_BVEventParser` is not a problem perse, but it breaks the conceptual structure.

File: mne/io/brainvision/tests/test_brainvision.py [hunk 7]
```diff
@@ -501,7 +501,7 @@ def test_read_vhdr_annotations_and_events():
     raw.annotations.append([1, 2, 3], 10, ['ZZZ', s_10, 'YYY'])
```
Comment (massich) [inline mne/io/brainvision/tests/test_brainvision.py]: The version from e64fd4fb7 simplifies `_check_event_id` and allow to remove `getattr` and the class methods. But requires to pass `event_id` because we are appending extra markers.

```diff
-    _, event_id = events_from_annotations(raw)
+    _, event_id = events_from_annotations(raw, event_id='brainvision')
```

If we want to keep the default, in this case, we need to keep the class method etc. Or relax
`_check_bv_annot` and as soon as we detect one 'brainvision' marker, parse everything as 'brainvision'

WDYT @agramfort @jona-sassenhagen? How would you like to write the call?

**Comment by massich** [comment]
I've made the imports relative but they can't go outside of the function.

File: mne/io/brainvision/tests/test_brainvision.py [hunk 8]
```diff
@@ -501,7 +501,7 @@ def test_read_vhdr_annotations_and_events():
...
     expected_event_id.update(YYY=10001, ZZZ=10002)  # others starting at 10001
     expected_event_id[s_10] = 10
-    _, event_id = events_from_annotations(raw)
+    _, event_id = events_from_annotations(raw, event_id='brainvision')
```
Comment (agramfort) [inline mne/io/brainvision/tests/test_brainvision.py]: is this change necessary? or is it just to have this option covered by a test?

just checking if there is any public API change.

File: mne/annotations.py (hunk 5 above)
Comment (agramfort) [reply mne/annotations.py]: you should be able to default to _BVEventParser() only if you have a RawBrainVision, RawFIF or RawArray

File: mne/io/brainvision/tests/test_brainvision.py (hunk 8 above)
Comment (massich) [reply mne/io/brainvision/tests/test_brainvision.py]: see https://github.com/mne-tools/mne-python/pull/6417#discussion_r290898838

File: mne/annotations.py (hunk 5 above)
Comment (massich) [reply mne/annotations.py]: If that's the contract then we need to create `RawFiF` which we don't have.
```py
class A:
    pass
class B(A):
    pass
a = A()
b = B()
print(isinstance(a, A), isinstance(b,A), issubclass(type(b),A), issubclass(type(a),A))
```
```sh
True True True True
```

**Comment by agramfort** [comment]
mne/viz/raw.py:        from ..io import RawFIF, RawArray
mne/viz/raw.py:        if not isinstance(raw, (RawFIF, RawArray)):

RawFIF is just link to Raw but this would work no?

>

**Comment by massich** [comment]
Not as it is here:
https://github.com/mne-tools/mne-python/blob/3f4c82a7e7401a38c485deca7d9e3b655d605b62/mne/io/__init__.py#L50

This is the problem of fif being the base object.

But we can always resurrect `RawFIF`

**Comment by massich** [comment]
Scratch anything I said. `RawBrainVision(BaseRaw)` not `Raw`. `Raw` and `RawBrainVision` they are indeed siblings.

**Comment by massich** [comment]
I think this makes it. Merge if everyone is happy.

**Comment by agramfort** [comment]
good enough for me !

@massich can you see why travis complains?

**Comment by massich** [comment]
I would say its not related. its a precision error in `lcmv`.

**Comment by agramfort** [comment]
does this happen in other PRs?

>

**Comment by massich** [comment]
I could not replicate in my machine. so I restarted Travis to see if its just random.

```sh
~/code/mne-python bv_events*
(mne35) ❯ pytest mne/beamformer                                                    
Test session starts (platform: linux, Python 3.5.5, pytest 3.8.1, pytest-sugar 0.9.2)
rootdir: /home/sik/code/mne-python, inifile: setup.cfg
plugins: sugar-0.9.2, faulthandler-1.5.0, cov-2.6.0

 mne/beamformer/tests/test_dics.py s✓✓✓✓✓✓✓✓✓✓                                                                                19% █▉        
 mne/beamformer/tests/test_external.py ✓✓✓✓                                                                                   26% ██▋       
 mne/beamformer/tests/test_lcmv.py ✓ssss✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓                                                   95% █████████▌
 mne/beamformer/tests/test_rap_music.py ✓✓✓                                                                                  100% ██████████
------------------------------------- generated xml file: /home/sik/code/mne-python/junit-results.xml --------------------------------------


========================================================= short test summary info ==========================================================
SKIP [1] mne/beamformer/tests/test_dics.py:107: Test test_make_dics skipped, requires h5py. Got exception (No module named 'h5py')
SKIP [4] mne/beamformer/tests/test_lcmv.py:199: Test test_make_lcmv skipped, requires h5py. Got exception (No module named 'h5py')
======================================================== slowest 20 test durations =========================================================
3.00s call     mne/beamformer/tests/test_lcmv.py::test_lcmv_vector
2.58s call     mne/beamformer/tests/test_lcmv.py::test_lcmv_reg_proj[unit-noise-gain-True]
2.54s call     mne/beamformer/tests/test_lcmv.py::test_lcmv_reg_proj[unit-noise-gain-False]
2.46s call     mne/beamformer/tests/test_dics.py::test_tf_dics[testing_data]
2.21s call     mne/beamformer/tests/test_lcmv.py::test_lcmv_reg_proj[nai-True]
2.03s call     mne/beamformer/tests/test_lcmv.py::test_lcmv_reg_proj[None-False]
2.00s call     mne/beamformer/tests/test_lcmv.py::test_lcmv_reg_proj[nai-False]
1.97s call     mne/beamformer/tests/test_lcmv.py::test_tf_lcmv
1.96s call     mne/beamformer/tests/test_lcmv.py::test_lcmv_reg_proj[None-True]
1.55s call     mne/beamformer/tests/test_dics.py::test_apply_dics_timeseries[testing_data]
1.46s call     mne/beamformer/tests/test_rap_music.py::test_rap_music_simulated
1.44s call     mne/beamformer/tests/test_rap_music.py::test_rap_music_sphere
1.34s call     mne/beamformer/tests/test_lcmv.py::test_lcmv_source_power
1.25s setup    mne/beamformer/tests/test_external.py::test_lcmv_fieldtrip[testing_data-ug_vec-None-None-False]
1.08s setup    mne/beamformer/tests/test_external.py::test_lcmv_fieldtrip[testing_data-ung-unit-noise-gain-max-power-False]
1.02s call     mne/beamformer/tests/test_lcmv.py::test_depth_does_not_matter[testing_data-testing_data-max-power-nai]
0.99s call     mne/beamformer/tests/test_lcmv.py::test_depth_does_not_matter[testing_data-testing_data-max-power-unit-noise-gain]
0.97s setup    mne/beamformer/tests/test_external.py::test_lcmv_fieldtrip[testing_data-ug_scal-None-max-power-False]
0.96s call     mne/beamformer/tests/test_lcmv.py::test_lcmv_ctf_comp
0.94s call     mne/beamformer/tests/test_lcmv.py::test_depth_does_not_matter[testing_data-testing_data-vector-nai]

Results (73.02s):
      53 passed
       5 skipped

```

**Comment by massich** [comment]
It passed here https://travis-ci.org/mne-tools/mne-python/jobs/542190809#L2724

**Comment by massich** [comment]
@jona-sassenhagen does this solve your usecase?

File: mne/io/brainvision/brainvision.py (hunk 4 above)
Comment (larsoner) [reply mne/io/brainvision/brainvision.py]: Ahh right `set(...).union(set(...))`

**Comment by larsoner** [comment]
@jona-sassenhagen feel free to merge if you're happy

File: mne/annotations.py [hunk 9]
```diff
@@ -817,10 +817,21 @@ def _select_annotations_based_on_description(descriptions, event_id, regexp):
...
+    from .io.brainvision.brainvision import RawBrainVision
+    from .io import RawFIF, RawArray
+
     if event_id is None:
         return _DefaultEventParser()
     elif event_id == 'auto':
-        return getattr(raw, '_get_auto_event_id', _DefaultEventParser)()
+        if isinstance(raw, RawBrainVision) or (
+            isinstance(raw, (RawFIF, RawArray)) and
+            _check_bv_annot(raw.annotations.description)
```
Comment (jona-sassenhagen) [inline mne/annotations.py]: Maybe we should log/warn in case the BV parser is called on a RawFIF/RawArray?

Comment (larsoner) [reply mne/annotations.py]: I would not warn because the save-load round-trip should be a reasonable use case. But a `logger.info` seems fine

File: mne/io/brainvision/brainvision.py [hunk 10]
```diff
@@ -849,3 +844,10 @@ def __call__(self, description):
             code = (super(_BVEventParser, self)
                     .__call__(description, offset=_OTHER_OFFSET))
         return code
+
+
+def _check_bv_annot(descriptions):
+    markers_basename = set([dd.rstrip('0123456789 ') for dd in descriptions])
+    bv_markers = (set(_BV_EVENT_IO_OFFSETS.keys())
+                  .union(set(_OTHER_ACCEPTED_MARKERS.keys())))
+    return len(markers_basename - bv_markers) == 0
```
Comment (jona-sassenhagen) [inline mne/io/brainvision/brainvision.py]: We could be even stricter here if I understand correctly that the numeric code ending a BV code is always 3 characters (whitespace left padded plus an int 0-255), i.e., 

`set([dd[:-3] for dd in descriptions])`

Comment (jona-sassenhagen) [reply mne/io/brainvision/brainvision.py]: @cbrnr this is the case right?

**Comment by jona-sassenhagen** [comment]
Thanks for doing this @massich . If you notify when you've addressed my two short comments, I would merge.

File: mne/io/brainvision/brainvision.py (hunk 10 above)
Comment (cbrnr) [reply mne/io/brainvision/brainvision.py]: I think so, at least I've only ever seen markers like that (`S  2`, `R  1`, ...).

Comment (massich) [reply mne/io/brainvision/brainvision.py]: I started by `set([dd[:-3] for dd in descriptions])` but I went for `rstrip` because I found it more readable. But either one is fine.

**Comment by massich** [comment]
I've also rebased. Feel free to merge.

**Comment by jona-sassenhagen** [comment]
The Travis builds are failing?

**Comment by larsoner** [comment]
Yes and the failure is real

https://travis-ci.org/mne-tools/mne-python/jobs/544263546

**Comment by massich** [comment]
yes I think that `sync` or some other had some extra space. `rstrip` should work

**Comment by larsoner** [comment]
Thanks @massich

**Comment by jona-sassenhagen** [comment]
Thanks @massich !

**Comment by jona-sassenhagen** [comment]
Oh wow, I found a file that's become unreadable with this PR!

```
    826         pat = re.compile(r'^(\s*def\s)|(\s*async\s+def\s)|(.*(?<!\w)lambda(:|\s))|^(\s*@)')
    827         while lnum > 0:
--> 828             if pat.match(lines[lnum]): break
    829             lnum = lnum - 1
    830         return lines, lnum

IndexError: list index out of range
```

I am investigating.
